File: php/tests/Unit/Advertisement/Domain/ValueObject/EmailTest.php
<?php
declare(strict_types=1);

namespace Tests\Demo\App\Unit\Advertisement\Domain\ValueObject;

use Demo\App\Advertisement\Domain\Exceptions\InvalidEmailException;
use Demo\App\Advertisement\Domain\Model\ValueObject\Email;
use PHPUnit\Framework\TestCase;
use ReflectionClass;

class EmailTest extends TestCase
{
    private const string EMAIL = 'email@test.com';
    private const string INVALID_EMAIL = 'emailtest.com';

    public function testShouldNotBeInstantiatedWithTheConstructor(): void
    {
        $class = new ReflectionClass(Email::class);
        $constructor = $class->getConstructor();

        self::assertTrue($constructor->isPrivate(), 'Constructor is not private');
    }

    public function testShouldCreateAnEmail()
    {
        $result = Email::build(self::EMAIL);

        $email = $result->unwrap();

        self::assertInstanceOf(Email::class, $email);
        self::assertEquals(self::EMAIL, $email->value());
    }

    public function testShouldReturnErrorResultWhenEmailIsInvalid()
    {
        $result = Email::build(self::INVALID_EMAIL);

        self::assertFalse($result->isSuccess());
        self::assertInstanceOf(InvalidEmailException::class, $result->exception());
        self::assertEquals('Invalid email format ' . self::INVALID_EMAIL, $result->exception()->getMessage());
    }
}

File: kotlin/src/test/kotlin/unit/advertisement/domain/model/value_object/EmailTest.kt
package unit.advertisement.domain.model.value_object

import advertisement.domain.model.value_object.Email
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.lang.reflect.Modifier

class EmailTest
{
    companion object {
        private const val VALID_EMAIL = "email@test.con"
        private const val INVALID_EMAIL = "email.test.con"
    }

    @Test
    fun testShouldNotBeInstantiatedWithTheConstructor() {
        Assertions.assertThrows(NoSuchMethodException::class.java) {
            Assertions.assertTrue(Modifier.isPrivate(Email::class.java.getDeclaredConstructor().modifiers))
        }
    }

    @Test
    fun testShouldCreateAnEmail() {
        val result = Email.build(VALID_EMAIL)

        result.fold(
            { error -> Assertions.fail("Expected a valid email, but got error: $error") },
            { id -> Assertions.assertEquals(VALID_EMAIL, id.value()) }
        )
    }

    @Test
    fun testShouldThrowAnExceptionWhenEmailIsInvalid() {
        val result = Email.build(INVALID_EMAIL)

        result.fold(
            { error ->
                Assertions.assertEquals("Invalid email format $INVALID_EMAIL", error.message)
            },
            { description -> Assertions.fail("Expected an error, but got a valid email: ${description.value()}") }
        )
    }
}
